### PR TITLE
Feature/pivotcholesky

### PR DESCRIPTION
**Pull request by edaub**
Implements pivoted Cholesky decomposition as a method for handling the emulator nugget. If this option is chosen, the Cholesky factorization uses pivoting to avoid problems with singular matrices and automatically ignores points that are numerically similar enough to others such that the covariance matrix is nearly singular. Addresses issue #9.

Changes in this PR:

* Implements pivoted Cholesky decomposition in the `linalg/cholesky.py` file, plus a function for inverting pivot matrices.
* Modifications to the `GaussianProcess` class to allow pivoting as an option through setting the `nugget` attribute.
* Additional changes were required to re-order the inputs/targets to use the cached factorized covariance matrix (as the rows/columns have been re-arranged due to pivoting). This is done using the pivot matrix and its transpose as indices where appropriate in all calls to the `cho_solve` function -- once to rearrange the RHS before solving using the pivot, and again after solving with the transpose to return to the ordering in the original data.
* Includes a benchmark to illustrate how pivoting improves performance on problems with duplicate input/target values.
* Updates to documentation to reflect the above changes.

### CODE REVIEW

**Comment by edaub** [comment]
To Do: `cho_solve` function combined with array reordering using the pivot should be wrapped into a function to improve legibility and ensure it hasn't been forgotten anywhere.

File: docs/benchmarks/benchmarks.rst [hunk 1]
```diff
@@ -10,12 +10,13 @@ Benchmarks
    tsunami
    mcmc_benchmark
    mice_benchmark
+   pivot_benchmark
    gkdr_benchmark
    histmatch_benchmark
 
 The code includes a series of benchmarks that illustrate various pieces of the implementation. Benchmarks
 can be run from the ``mogp_emulator/tests`` directory by entering ``make all`` or ``make benchmarks`` to
```
Comment (nbarlowATI) [inline docs/benchmarks/benchmarks.rst:18]: I think this should nowadays be the ``mogp_emulator/benchmarks`` directory.

File: docs/benchmarks/benchmarks.rst [hunk 2]
```diff
@@ -62,6 +63,14 @@ This creates designs of a variety of sizes and computes the error on unseen data
 function. It compares the accuracy of the sequential design to the Latin Hypercube for both the
 predictions and uncertainties.
 
+Pivoted Cholesky Benchmark
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A benchmark shows how a duplicated input can degrade emulator performance
```
Comment (nbarlowATI) [inline docs/benchmarks/benchmarks.rst:69]: "A benchmark shows how ..." => "A benchmark that shows how ..." ?

**Review by nbarlowATI — APPROVED** [review]
Looks good!  The pivot benchmark, and all the tests work for me.
I guess some linting was done to reduce the line lengths on the docstrings in GaussianProcess.py which makes the diff look huge, but the actual changes to the code and docstrings all looked fine to me.

File: docs/benchmarks/benchmarks.rst (hunk 1 above)
Comment (edaub) [reply docs/benchmarks/benchmarks.rst:18]: Ah, yes, good catch. I have a separate change ready to go that re-does this file, so I'll fix this page up when I push those changes.

File: docs/benchmarks/benchmarks.rst (hunk 2 above)
Comment (edaub) [reply docs/benchmarks/benchmarks.rst:69]: Also good catch, as with the previous I'll fix with another set of changes I have ready to go.

**Pull request merged**